File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = 201
    
    leaf = deque()
    zero = [0] * (n+1)
    graph = [[] for _ in range(n+1)]
    visit = [False] * (n+1)
    cnt = [0] * (n+1)
    
    for w in wires :
        a, b = w[0], w[1]
        zero[a] += 1
        zero[b] += 1
        graph[a].append(b)
        graph[b].append(a)
        
    for i in range(1,n+1) :
        if zero[i] == 1 :
            leaf.append(i)
    
    while leaf :
        current_node = leaf.popleft()
        print("current_node: ", current_node)
        
        tmp = 0
        for nodes in graph[current_node] :
            if not visit[nodes] :
                next_node = nodes
                tmp = 1
                break
                
        if tmp == 0 :
            break
        
        print("next_node: ", next_node)
        next_node_cnt = n - (cnt[current_node]+1)
        print("current_cnt: ", cnt[current_node]+1)
        print("next_cnt: ", next_node_cnt)
        answer = min(answer, abs(next_node_cnt - cnt[current_node] -1))
        zero[next_node] -= 1
        cnt[next_node] += cnt[current_node] + 1
        
        if zero[next_node] == 1 :
            leaf.append(next_node)

        visit[current_node] = True
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    int answer;
    
    public int solution(int n, int[][] wires) {
        answer = n+1;
        
        List<int[]> wires_list = new ArrayList<>();
        for (int[] wire : wires) {
            wires_list.add(wire);
        }
        
        int m = wires.length;
        for (int i=0; i<m; i++) {
            int[] tmp = wires[i];
            wires_list.remove(i);
            // System.out.println();
            // for (int[] tt : wires_list) {
            //     System.out.print(tt[0]);
            //     System.out.print(" ");
            //     System.out.print(tt[1]);
            //     System.out.print(" ");
            // }
            bfs(wires_list, n);
            wires_list.add(i, tmp);
        }
        
        return answer;
    }
    
    public void bfs(List<int[]> wires, int n) {
        boolean[] visit = new boolean[n+1];
        List<List<Integer>> graph = new ArrayList<>();
        
        for (int i=0; i<n+1; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int[] w : wires) {
            int a = w[0];
            int b = w[1];
            
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        int cnt = 0;
        
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(1);
        cnt++;
        visit[1] = true;
        
        while (!queue.isEmpty()) {
            int value = queue.poll();
            
            for (int neighbor : graph.get(value)) {
                if (visit[neighbor]) {
                    continue;
                }
                queue.offer(neighbor);
                cnt++;
                visit[neighbor] = true;
            }
        }
        
        int tmp = Math.abs(cnt - (n - cnt));
        answer = Math.min(answer, tmp);
    }
    
}